File: src/main/java/org/athenian/Palindrome.java
package org.athenian;

import java.util.stream.IntStream;

public class Palindrome {

    public static boolean isPalindromeLoop(String s) {
        int len = s.length();
        for (int i = 0; i < s.length() / 2; i++) {
            String beginLetter = s.substring(i, i + 1);
            String endLetter = s.substring(len - i - 1, len - i);
            if (!beginLetter.equals(endLetter))
                return false;
        }
        return true;
    }

    public static boolean isPalindromeRecursion(String s) {
        if (s.length() <= 1)
            return true;

        String firstLetter = s.substring(0, 1);
        String lastLetter = s.substring(s.length() - 1);
        if (!firstLetter.equals(lastLetter))
            return false;

        String middleLetters = s.substring(1, s.length() - 1);
        return isPalindromeRecursion(middleLetters);
    }

    public static boolean isPalindromeStream(String s) {
        char[] str = s.toCharArray();
        int len = str.length;

        return IntStream
                .range(0, len / 2)
                .allMatch(i -> str[i] == (str[len - i - 1]));
    }

}

File: src/main/kotlin/org/athenian/Palindrome.kt
package org.athenian

fun isPalindromeLoop(s: String) =
    when {
        s.isEmpty() -> true
        else -> {
            val len = s.length
            repeat(len / 2) {
                val beginLetter = s[it]
                val endLetter = s[len - it - 1]
                if (beginLetter != endLetter)
                    return false
            }
            true
        }
    }

fun isPalindromeRecursion(s: String): Boolean =
    when {
        s.length <= 1 -> true
        s[0] != s[s.length - 1] -> false
        else -> isPalindromeRecursion(s.slice(1..s.length - 2))
    }

fun isPalindromeStream(s: String) =
    when {
        s.length <= 1 -> true
        else -> {
            val half = s.length / 2
            val firstHalf = s.slice(0..half - 1)
            val secondHalf = s.slice(s.length - half..s.length - 1)
            firstHalf
                .zip(secondHalf.reversed())
                .all { (c1, c2) -> c1 == c2 }
        }
    }